File: 19_Remove_Nth_Node_From_End_of_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy= new ListNode(-1);
        ListNode* p1 = dummy;
        ListNode* p2 = dummy;
        dummy->next = head;
        for (int i=0;i<n;i++){
            p1=p1->next;
        }
        // if (p1==nullptr){
        //     return head;
        // }
        while(p1->next!=nullptr){
            p1=p1->next;
            p2=p2->next;
        }
        // p2->next is last n th node
        p2->next = p2->next->next;
        return dummy->next;
    }
};

File: 19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy= new ListNode(-1);
        ListNode p1 = dummy, p2 = dummy;
        dummy.next = head;
        for (int i = 0; i < n; i++){
            p1 = p1.next;
        }
        while(p1.next != null){
            p1 = p1.next;
            p2 = p2.next;
        }
        // p2.next is last n th node
        p2.next = p2.next.next;
        return dummy.next;
    }
}
